Add Custom Script Engine

BalestraPatrick opened the pull request:
I continued working on #148 and fixed the conflicts with master that arose from the latest release, so I am creating a new PR for an easier review directly with master. Let me know if you prefer otherwise and I have another branch that doesn't have the conflicts fixed yet, so I could make a PR to `feature/engines` first and then to master. 👍 

I did some extensive testing by even prototyping a simply NodeJS server and a script that does the communication part taking in `list`, `upload` and `download` commands and it's working as expected.

I also threw in some random typos and documentation fixes noticed in reading the existing code.
I didn't run Brittany yet, so I am sure there will be a ton of fixes done to my bad Haskell 😄 

Regarding integration tests, I wrote `engine.sh` that quickly simulates an engine that simply stores the artifacts in a local folder. Added a bunch of integration tests covering all cases:
- `current-framework-engine-yml.bats`
- `current-framework-named-engine-yml.bats`
- `dynamic-frameworks-engine-yml.bats`
- `static-frameworks-engine-yml.bats`

I also renamed `current-framework-yaml.bats` to `current-framework-yml.bats` and `current-framework-named-yaml.bats` to `current-framework-named-yml.bats` for consistency.

Fixes #146

BalestraPatrick commented on src/Lib.hs:
```diff
@@ -315,8 +326,15 @@ runUDCCommand command absoluteRomefilePath verbose romeVersion = do
       <> romeVersionToString vers
       <> noColorControlSequence
 
+  -- case (mS3BucketName, mEnginePath) of
```
This is not needed at all I think, it was probably an initial sketch leftover?

blender replied:
yep, not needed. I was mapping all possible combinations.

blender commented on src/Engine/Probing.hs:
```diff
@@ -0,0 +1,71 @@
+    case exitCode of
+        -- If script exits with success, we assume the framework exists.
+        Turtle.ExitSuccess   -> return (PlatformAvailability platform True)
+        Turtle.ExitFailure n -> return (PlatformAvailability platform False)
```
```suggestion
        Turtle.ExitFailure _ -> return (PlatformAvailability platform False)
```

blender commented on src/Lib.hs:
```diff
@@ -1036,7 +1299,7 @@ downloadFrameworksAndArtifactsFromCaches s3BucketName mlCacheDir reverseRomeMap
 -- | Downloads a Framework and it's relative dSYM from and S3 Bucket or a local cache.
 -- | If the Framework and dSYM are not found in the local cache then they are downloaded from S3.
--- | If SkipLocalCache is specified, th local cache is ignored.
+-- | If SkipLocalCache is specified, the local cache is ignored.
```
👍

blender commented on src/Lib.hs:
```diff
@@ -1221,6 +1482,304 @@ downloadFrameworkAndArtifactsFromCaches s3BucketName Nothing reverseRomeMap fVer
+                                                                reverseRomeMap
+                                                                fVersion
+                                                              platform
+          flip whenLeft eitherSymbolmapsOrErrors $ \e -> case e of
```
nope

BalestraPatrick commented:
@blender PR is now green since I fixed the integration the tests and a crash with INI configurations in cd1eaa5e5c0a1b22a4580ef67ff2c0a3317d9b92. I added some information to the README in 9c453c987e88a321f05df8b27b12d6d4ec9b3fe6 since I wanted to document how it works, we can always adjust later when a new version is cut.

blender commented on README.md:
```diff
@@ -9,6 +9,7 @@ Rome is a tool that allows developers on Apple platforms to use:
 - [Ceph](https://ceph.com/ceph-storage/object-storage/)
 - other S3 compatible object stores
 - or/and a local folder
+- your own custom engine
```
```suggestion
- [your own custom engine](#customengine)
```

blender replied:
Let's market it :D

blender commented on README.md:
```diff
@@ -257,6 +259,25 @@ Default port for `http` endpoints is __9000__ if the port is left unspecified.
 
 Alternatively the endpoint can also be specified by setting an `AWS_ENDPOINT` environment variable.
 
+### Custom Engine
```
Great job adding this explanation to the readme

BalestraPatrick replied:
Thanks! 😄

blender commented:
PR looks good to me! 
Excellent job and amazing contribution! 🎉

thii commented:
@blender Could you release this to CocoaPods?

blender commented:
Sure! @thii there is a pre-release version attached to the releases.  Did you have a chance to try it?

thii commented:
I've tried it and found no problems so far.

blender commented:
@thii it's out
